luigi_pipeline/sex_related_checks.py

```diff
@@ -0,0 +1,568 @@
+    )
+    xx_fstat_threshold = luigi.FloatParameter(
+        default=0.50,
+        description="F-stat threshold below which a sample will be called XX. Default is 0.50.",
+    )
+    aaf_threshold = luigi.FloatParameter(
+        default=0.05,
+        description="Alternate allele frequency threshold for `hl.impute_sex`. Default is 0.05.",
+    )
+    callrate_threshold = luigi.FloatParameter(
+        default=0.25,
+        description="Minimum variant call rate threshold. Default is 0.25.",
+    )
+    normalization_contig = luigi.Parameter(
+        default="20",
+        description="Autosome to use to normalize sex chromosome coverage. Default is chromosome 20.",
+    )
+    hail_temp_dir = luigi.OptionalParameter(
+        default="gs://seqr-scratch-temp/",
+        description="Networked temporary directory used by hail for temporary file storage. Must be a network-visible file path.",
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def output_directory(self):
+        return GCSorLocalTarget(self.output_dir)
+
+    def temp_directory(self):
+        return GCSorLocalTarget(self.temp_dir)
+
+    def output(self):
+        return GCSorLocalTarget(f"{self.output_directory().path}sex_ploidy_check.ht")
+
+    def complete(self):
+        # Complete is called by Luigi to check if the task is done and will skip if it is.
+        # By default it checks to see that the output exists, but we want to check for the
+        # _SUCCESS file to make sure it was not terminated halfway.
+        return GCSorLocalTarget(f"{self.output().path}/_SUCCESS").exists()
+
+    def mt_path(self):
+        return GCSorLocalTarget(f"{self.output_directory().path}callset.mt")
+
+    def run(self):
+        # first validate paths
+        if self.hail_temp_dir:
+            hl.init(
+                tmp_dir=self.hail_temp_dir,
+            )  # Need to use the GCP bucket as temp storage for very large callset joins
+        hl._set_flags(
+            use_new_shuffle="1"
+        )  # Interval ref data join causes shuffle death, this prevents it
+
+        for source_path in [self.source_paths]:
+            check_if_path_exists(source_path, "source_path")
+        if self.wes_filter_source_paths:
+            for wes_filter_source_path in self.wes_filter_source_paths:
+                check_if_path_exists(source_path, "source_path")
+        if self.hail_temp_dir:
+            check_if_path_exists(self.hail_temp_dir, "hail_temp_dir")
+
+        self.read_input_write_mt()
+        self.call_sex()
+
+    def call_sex(
+        self,
+        use_y_cov: bool = False,
+        add_x_cov: bool = False,
+        y_cov_threshold: float = 0.1,
+        normalization_contig: str = "20",
+        xy_fstat_threshold: float = 0.75,
+        xx_fstat_threshold: float = 0.5,
+        aaf_threshold: float = 0.05,
+        call_rate_threshold: float = 0.25,
+        final_annotations: List[str] = [
```

Comment:
Aha!  This is a [common python gotcha](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments).

I think maybe the thing to do here is to move this list to the top of the file as a constant:

```
FINAL_ANNOTATIONS = [...]
```

and then inside this function make use of that constant:

```
if use_y_cov:
    final_annotations = [*FINAL_ANNOTATIONS, *[
                     f"chr{normalization_contig}_mean_dp",
                     "chrY_mean_dp",
                     "normalized_y_coverage",
                 ]]
```

(In reality the mutability is unlikely to be a bug for us because we only invoke these functions once time per task run and then kill the dataproc cluster).